File: app/src/modules/submissions/dtos/submission-with-membership-response-dto.ts
import {FormStatus} from '../../forms/enums/form-status';
import {FormType} from '../../forms/enums/form-type';

export interface SubmissionWithMembershipResponseDTO {
    id: string;
    formId: number | null;
    created: string; // ISO date string
    assigneeId: string | null;
    archived: string | null; // ISO date string
    fileNumber: string | null;
    destinationId: number | null;
    customerInput: any; // ElementData type
    destinationSuccess: boolean | null;
    isTestSubmission: boolean | null;
    copySent: boolean | null;
    copyTries: number | null;
    reviewScore: number | null;
    destinationResult: string | null;
    destinationTimestamp: string | null; // ISO date string
    status: string; // SubmissionStatus enum
    updated: string; // ISO date string
    paymentTransactionKey: string | null;

    formVersion: number | null;
    formSlug: string | null;

    formInternalTitle: string | null;
    formPublicTitle: string | null;

    formDevelopingDepartmentId: number | null;
    formManagingDepartmentId: number | null;
    formResponsibleDepartmentId: number | null;

    formCreated: string | null; // ISO date string
    formUpdated: string | null; // ISO date string

    formPublishedVersion: number | null;
    formDraftedVersion: number | null;

    formStatus: FormStatus; // FormStatus enum
    formType: FormType; // FormType enum

    formLegalSupportDepartmentId: number | null;
    formTechnicalSupportDepartmentId: number | null;
    formImprintDepartmentId: number | null;
    formPrivacyDepartmentId: number | null;
    formAccessibilityDepartmentId: number | null;

    formCustomerAccessHours: number | null;
    formSubmissionRetentionWeeks: number | null;

    formThemeId: number | null;

    formPdfTemplateKey: string | null; // UUID as string

    formPaymentProviderKey: string | null; // UUID as string
    formPaymentPurpose: string | null;
    formPaymentDescription: string | null;
    formPaymentProducts: any[]; // PaymentProduct[]

    formIdentityProviders: any[]; // IdentityProviderLink[]
    formIdentityVerificationRequired: boolean | null;

    formDestinationId: number | null;

    formRootElement: any; // RootElement

    formVersionPublished: string | null; // ISO date string
    formVersionRevoked: string | null; // ISO date string

    userId: string | null;
    userEmail: string | null;
    userFirstName: string | null;
    userLastName: string | null;
    userFullName: string | null;
    userEnabled: boolean | null;
    userVerified: boolean | null;
    userGlobalAdmin: boolean | null;
    userDeletedInIdp: boolean | null;
    userIsDeveloper: boolean | null;
    userIsManager: boolean | null;
    userIsResponsible: boolean | null;
}

File: src/main/java/de/aivot/GoverBackend/submission/dtos/SubmissionWithMembershipResponseDTO.java
package de.aivot.GoverBackend.submission.dtos;

import de.aivot.GoverBackend.elements.models.ElementData;
import de.aivot.GoverBackend.elements.models.elements.RootElement;
import de.aivot.GoverBackend.enums.SubmissionStatus;
import de.aivot.GoverBackend.form.enums.FormStatus;
import de.aivot.GoverBackend.form.enums.FormType;
import de.aivot.GoverBackend.identity.models.IdentityProviderLink;
import de.aivot.GoverBackend.models.payment.PaymentProduct;
import de.aivot.GoverBackend.submission.entities.SubmissionWithMembership;

import java.time.LocalDateTime;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

public record SubmissionWithMembershipResponseDTO(
        String id,
        Integer formId,
        LocalDateTime created,
        String assigneeId,
        LocalDateTime archived,
        String fileNumber,
        Integer destinationId,
        ElementData customerInput,
        Boolean destinationSuccess,
        Boolean isTestSubmission,
        Boolean copySent,
        Integer copyTries,
        Integer reviewScore,
        String destinationResult,
        LocalDateTime destinationTimestamp,
        SubmissionStatus status,
        LocalDateTime updated,
        String paymentTransactionKey,

        Integer formVersion,
        String formSlug,

        String formInternalTitle,
        String formPublicTitle,

        Integer formDevelopingDepartmentId,
        Integer formManagingDepartmentId,
        Integer formResponsibleDepartmentId,

        LocalDateTime formCreated,
        LocalDateTime formUpdated,

        Integer formPublishedVersion,
        Integer formDraftedVersion,

        FormStatus formStatus,
        FormType formType,

        Integer formLegalSupportDepartmentId,
        Integer formTechnicalSupportDepartmentId,
        Integer formImprintDepartmentId,
        Integer formPrivacyDepartmentId,
        Integer formAccessibilityDepartmentId,

        Integer formCustomerAccessHours,
        Integer formSubmissionRetentionWeeks,

        Integer formThemeId,

        UUID formPdfTemplateKey,

        UUID formPaymentProviderKey,
        String formPaymentPurpose,
        String formPaymentDescription,
        Collection<PaymentProduct> formPaymentProducts,

        List<IdentityProviderLink> formIdentityProviders,
        Boolean formIdentityVerificationRequired,

        Integer formDestinationId,

        RootElement formRootElement,

        LocalDateTime formVersionPublished,
        LocalDateTime formVersionRevoked,

        String userId,
        String userEmail,
        String userFirstName,
        String userLastName,
        String userFullName,
        Boolean userEnabled,
        Boolean userVerified,
        Boolean userGlobalAdmin,
        Boolean userDeletedInIdp,
        Boolean userIsDeveloper,
        Boolean userIsManager,
        Boolean userIsResponsible
) {

    public static SubmissionWithMembershipResponseDTO from(SubmissionWithMembership submission) {

        return new SubmissionWithMembershipResponseDTO(
                submission.getId(),
                submission.getFormId(),
                submission.getCreated(),
                submission.getAssigneeId(),
                submission.getArchived(),
                submission.getFileNumber(),
                submission.getDestinationId(),
                submission.getCustomerInput(),
                submission.getDestinationSuccess(),
                submission.getTestSubmission(),
                submission.getCopySent(),
                submission.getCopyTries(),
                submission.getReviewScore(),
                submission.getDestinationResult(),
                submission.getDestinationTimestamp(),
                submission.getStatus(),
                submission.getUpdated(),
                submission.getPaymentTransactionKey(),

                submission.getFormVersion(),
                submission.getFormSlug(),

                submission.getFormInternalTitle(),
                submission.getFormPublicTitle(),

                submission.getFormDevelopingDepartmentId(),
                submission.getFormManagingDepartmentId(),
                submission.getFormResponsibleDepartmentId(),

                submission.getFormCreated(),
                submission.getFormUpdated(),

                submission.getFormPublishedVersion(),
                submission.getFormDraftedVersion(),

                submission.getFormStatus(),
                submission.getFormType(),

                submission.getFormLegalSupportDepartmentId(),
                submission.getFormTechnicalSupportDepartmentId(),
                submission.getFormImprintDepartmentId(),
                submission.getFormPrivacyDepartmentId(),
                submission.getFormAccessibilityDepartmentId(),

                submission.getFormCustomerAccessHours(),
                submission.getFormSubmissionRetentionWeeks(),

                submission.getFormThemeId(),

                submission.getFormPdfTemplateKey(),

                submission.getFormPaymentProviderKey(),
                submission.getFormPaymentPurpose(),
                submission.getFormPaymentDescription(),
                submission.getFormPaymentProducts(),

                submission.getFormIdentityProviders(),
                submission.getFormIdentityVerificationRequired(),

                submission.getFormDestinationId(),

                submission.getFormRootElement(),

                submission.getFormVersionPublished(),
                submission.getFormVersionRevoked(),

                submission.getUserId(),
                submission.getUserEmail(),
                submission.getUserFirstName(),
                submission.getUserLastName(),
                submission.getUserFullName(),
                submission.getUserEnabled(),
                submission.getUserVerified(),
                submission.getUserGlobalAdmin(),
                submission.getUserDeletedInIdp(),
                submission.getUserIsDeveloper(),
                submission.getUserIsManager(),
                submission.getUserIsResponsible()
        );
    }

}
